pkg/object/gatewaycontroller/translator.go

```diff
@@ -0,0 +1,660 @@
+	if b.reqAdaptor == nil {
+		b.reqAdaptor = &builder.RequestAdaptorSpec{}
+	}
+
+	header := &httpheader.AdaptSpec{
+		Add: map[string]string{},
+		Set: map[string]string{},
+		Del: []string{},
+	}
+	for i := range f.Add {
+		h := &f.Add[i]
+		header.Add[string(h.Name)] = h.Value
+	}
+	for i := range f.Set {
+		h := &f.Set[i]
+		header.Set[string(h.Name)] = h.Value
+	}
+	header.Del = f.Remove
+
+	b.reqAdaptor.Header = header
+}
+
+func (b *pipelineSpecBuilder) addURLRewrite(f *gwapis.HTTPURLRewriteFilter) {
+	if b.reqAdaptor == nil {
+		b.reqAdaptor = &builder.RequestAdaptorSpec{}
+	}
+
+	if f.Hostname != nil {
+		b.reqAdaptor.Host = string(*f.Hostname)
+	}
+
+	if f.Path == nil {
+		return
+	}
+
+	if f.Path.Type == gwapis.FullPathHTTPPathModifier {
+		b.reqAdaptor.Path = &pathadaptor.Spec{Replace: *f.Path.ReplaceFullPath}
+	} else {
+		b.reqAdaptor.Path = &pathadaptor.Spec{TrimPrefix: *f.Path.ReplacePrefixMatch}
+	}
+}
+
+func (b *pipelineSpecBuilder) addRequestRedirect(f *gwapis.HTTPRequestRedirectFilter) {
+	// TODO: The current redirector filter does not compatible with the
+	// Gateway API spec.
+	logger.Errorf("redirector filter is not supported currently")
+	/*
+		if b.redirector == nil {
+			b.redirector = &redirector.Spec{StatusCode: 302}
+		}
+
+		var repl string
+
+		re := `^([^:]*)://([^:/]+)(:\d+)?/`
+		if f.Scheme == nil {
+			repl = "$1://"
+		} else {
+			repl = *f.Scheme + "://"
+		}
+
+		if f.Hostname == nil {
+			repl += "$2"
+		} else {
+			repl += string(*f.Hostname)
+		}
+
+		if f.Port == nil {
+			repl += "$3"
+		} else {
+			repl += fmt.Sprintf(":%d", *f.Port)
+		}
+
+		if f.StatusCode != nil {
+			b.redirector.StatusCode = *f.StatusCode
+		}
+	*/
+}
+
+func (b *pipelineSpecBuilder) addRequestMirror(addr string) {
+	if b.proxy == nil {
+		b.proxy = &httpproxy.Spec{}
+	}
+
+	spec := &httpproxy.ServerPoolSpec{
+		Filter: &httpproxy.RequestMatcherSpec{
+			RequestMatcherBaseSpec: proxies.RequestMatcherBaseSpec{
+				Policy: "random",
+				Permil: 1000, // set to 1000 to send all requests to the mirror
+			},
+		},
+		ServerMaxBodySize: -1,
+	}
+	spec.Servers = append(spec.Servers, &proxies.Server{URL: addr})
+
+	b.proxy.Pools = append(b.proxy.Pools, spec)
+}
+
+func (b *pipelineSpecBuilder) addResponseHeaderModifier(f *gwapis.HTTPHeaderFilter) {
+	if b.respAdaptor == nil {
+		b.respAdaptor = &builder.ResponseAdaptorSpec{}
+	}
+
+	header := &httpheader.AdaptSpec{
+		Add: map[string]string{},
+		Set: map[string]string{},
+		Del: []string{},
+	}
+	for i := range f.Add {
+		h := &f.Add[i]
+		header.Add[string(h.Name)] = h.Value
+	}
+	for i := range f.Set {
+		h := &f.Set[i]
+		header.Set[string(h.Name)] = h.Value
+	}
+	header.Del = f.Remove
+
+	b.respAdaptor.Header = header
+}
+
+func (b *pipelineSpecBuilder) addBackend(addr string, weight int) {
+	if b.proxy == nil {
+		b.proxy = &httpproxy.Spec{}
+		b.proxy.Pools = append(b.proxy.Pools, &httpproxy.ServerPoolSpec{ServerMaxBodySize: -1})
```

Comment:
A side question, why not make the default value 0 for unlimited body?

Reply:
To support unlimited body, we have to enable the [stream](https://github.com/megaease/easegress/blob/main/doc/reference/stream.md) mode, but in this mode, some of the new features of v2.0 becomes unavailable. so the meaning of "set `ServerMaxBodySize` to zero" is a balance of different things, and we think using `-1` for unlimited body is better for more cases.